### PR TITLE
Feature - Improve sockets #3

### PR DESCRIPTION
**Pull request by zhuyuanmao**
Improved the security and efficiency of sockets.  #3
* Changed `online_users` from `array` to `set` in python so the searching will take O(1) instead of O(n). 
* Added `jwt` token authentication when the socket client wants to connect to the server. 
* Disabled sockets auto connection on the client sides, only connect to the server when clients are authenticated.

### CODE REVIEW

File: server/online_users.py [hunk 1]
```diff
@@ -1 +1 @@
-online_users = []
+online_users = set()
```
Comment (FilipeGorodscy) [inline server/online_users.py:1]: 💯

File: server/socketio_app/views.py [hunk 2]
```diff
@@ -10,14 +14,23 @@
...
+def connect(sid, environ, auth):
+    token = auth.get("token", None)
+    user = None
+    try:
+        decoded = jwt.decode(token, SECRET_KEY, algorithms=["HS256"])
+        user = User.get_by_id(decoded["id"])
+    except (jwt.ExpiredSignatureError, jwt.InvalidSignatureError, jwt.DecodeError):
+        raise Exception("Authentication failed.")
+    if not user:
+        raise Exception("Can not find the user.")
```
Comment (FilipeGorodscy) [inline server/socketio_app/views.py]: 💯

**Review by FilipeGorodscy — COMMENTED** [review]
Great job overall! There are a few other things we'd like to see:

- We actually want the socket to be closed at some point.

Please let us know on Slack when this PR is ready for a second review!

**Comment by zhuyuanmao** [comment]
* Changed broadcast messaging to private messaging.
* Disconnect client when authentication failed or logout.

File: client/src/socket.js [hunk 3]
```diff
@@ -23,7 +23,11 @@ socket.on("connect", () => {
     store.dispatch(setNewMessage(data.message, data.sender));
   });
   socket.on("read-conversation", (data) => {
-    store.dispatch(realAllMessages(data.conversationId,data.lastReadMessageId));
+    const {conversationId, lastReadMessageId, recipientId} = data
+    store.dispatch(realAllMessages(conversationId,lastReadMessageId,recipientId));
+  });
+  socket.on('connect_error', (e) => {
+    console.log(e.message);
```
Comment (0916dhkim) [inline client/src/socket.js]: Please use `console.error` for error messages.

File: server/socketio_app/views.py [hunk 4]
```diff
@@ -10,14 +14,27 @@
...
 @sio.event
-def connect(sid, environ):
+def connect(sid, environ, auth):
+    token = auth.get("token", None)
+    user = None
+    try:
+        decoded = jwt.decode(token, SECRET_KEY, algorithms=["HS256"])
+        user = User.get_by_id(decoded["id"])
+    except (jwt.ExpiredSignatureError, jwt.InvalidSignatureError, jwt.DecodeError):
+        sio.disconnect(sid)
```
Comment (0916dhkim) [inline server/socketio_app/views.py:24]: Great :100:

**Review by 0916dhkim — APPROVED** [review]
@zhuyuanmao Looks good to me. This PR is approved :+1: 
You can change `console.log` to `console.error` and merge this branch yourself.

**Pull request merged**